### ISSUE TITLE
startFuture() does not complete but flags return fine

### ISSUE THREAD

**Issue by panghy** [launchdarkly/flutter-client-sdk#48]
**Describe the bug**
In the latest build (2.0.2) on Android (emulator or device), calling `LDClient.startWithContext()` returns a future that resolves properly but then calling `LDClient.startFuture()` does not. The future does not resolve and `LDClient.isInitialized()` returns false (if a timeLimit was given, say 5s). However, this does not actually impact using flags as we can still requests for flag values.

**To reproduce**
1. Initialize with `LDClient.startWithContext()`
2. `LDClient.startFuture(timeLimit: const Duration(seconds: 5))`
3. Check `LDClient.isInitialized()` (it returns false)
4. But flags can load fine.

**Expected behavior**
Expects the future returned from `startFuture()` to complete normally and that `LDClient.isInitialize()` returns true.

**Logs**
```
print('Initializing LaunchDarklyDelegate');
LDConfig config = LDConfigBuilder(mobileKey).build();
LDContextBuilder contextBuilder = LDContextBuilder();
contextBuilder.kind('user', 'ldd-init');
print('Starting LaunchDarklyDelegate');
await LDClient.startWithContext(config, contextBuilder.build());
print('Waiting for LaunchDarklyDelegate to receive flags');
await LDClient.startFuture(timeLimit: const Duration(seconds: 5));
print('LaunchDarklyDelegate initialized: ${LDClient.isInitialized()}');
```

yields:
```
I/flutter (14272): Initializing LaunchDarklyDelegate
I/flutter (14272): Starting LaunchDarklyDelegate
I/flutter (14272): Waiting for LaunchDarklyDelegate to receive flags
I/flutter (14272): LaunchDarklyDelegate initialized: false
```

**SDK version**
2.0.2

**Language version, developer tools**
Flutter 3.7.12 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 4d9e56e694 (3 months ago) • 2023-04-17 21:47:46 -0400
Engine • revision 1a65d409c7
Tools • Dart 2.19.6 • DevTools 2.20.1

**OS/platform**
Android API 32

**Comment by tanderson-ld** [comment]
Thank you for this report!  We have filed this internally as 210467 and will get back to you when we have finished investigating.

**Comment by andrewescutia** [comment]
we are also seeing that if the following calls are made in quick succession we are receiving an error. The documentation makes it appear as though this should be functional without a `startFuture` call in between them. And was fine in previous versions. Android appears impacted. iOS is fine

```
 await LDClient.startWithContext(configuration, ldContext);
 await LDClient.identifyWithContext(ldContext);
```

error:
`LDClient.get() was called before init()!`

**Comment by Mik77o** [comment]
> we are also seeing that if the following calls are made in quick succession we are receiving an error. The documentation makes it appear as though this should be functional without a `startFuture` call in between them. And was fine in previous versions. Android appears impacted. iOS is fine
> 
> ```
>  await LDClient.startWithContext(configuration, ldContext);
>  await LDClient.identifyWithContext(ldContext);
> ```
> 
> error: `LDClient.get() was called before init()!`

My team also encounters exactly the same error on Android and we are afraid to update to the latest version. We've had several app crashes because of this bug.

**Comment by Mik77o** [comment]
We encounter this problem when launching the application the first time after upgrading from `launchdarkly_flutter_client_sdk: ^1.3.0` to `launchdarkly_flutter_client_sdk: ^2.0.2`.

Our logs:
_E/DartMessenger( 4274): com.launchdarkly.sdk.android.LaunchDarklyException: LDClient.get() was called before init()!
E/DartMessenger( 4274): at com.launchdarkly.sdk.android.LDClient.get(LDClient.java:279)_

**Comment by tanderson-ld** [comment]
Hello everyone.  I am investigating this issue and I information and questions for you.

@panghy :

1. I have been unable to reproduce this issue using the code snippet from your original post.  Here are the logs I get when I run that snippet.  Is it possible that this device was offline when you ran that code?  The flags may still be able to be evaluated as the SDK caches data for the last resolved context.  We did see a similar issue in 2.0.1, but was fixed in 2.0.2, so perhaps there is a deeper issue we aren't aware of.  Again, we appreciate you reporting what you're seeing, it helps us make a better product for everyon!

```
Emulator
11:52:20.427306 Starting LaunchDarklyDelegate
11:52:22.687531 Waiting for LaunchDarklyDelegate to receive flags
11:52:22.702975 LaunchDarklyDelegate initialized: true

Samsung S22
 11:58:46.609935 Starting LaunchDarklyDelegate
11:58:47.094041 Waiting for LaunchDarklyDelegate to receive flags
11:58:47.246395 LaunchDarklyDelegate initialized: true
```

**Comment by tanderson-ld** [comment]
@andrewescutia , can you verify the version of the SDK you saw this behavior on?  We had a bug on 2.0.1 that looked exactly like this, but should have been fixed in 2.0.2.  Perhaps we missed a case.  Also, when you say "quick succession", do you have elapsed time data to help us reproduce this?  Thank you again for responding!

**Comment by tanderson-ld** [comment]
@Mik77o, do you see this 100% of the time in the version migration case?  This is a big deal to us, we appreciate you reporting it and I will be looking into this one immediately.

**Comment by andrewescutia** [comment]
> @andrewescutia , can you verify the version of the SDK you saw this behavior on? We had a bug on 2.0.1 that looked exactly like this, but should have been fixed in 2.0.2. Perhaps we missed a case. Also, when you say "quick succession", do you have elapsed time data to help us reproduce this? Thank you again for responding!

Latest version appears to correct our issue. Thanks!

**Comment by Mik77o** [comment]
@tanderson-ld No, this error occurs only sometimes. It's not like we get this error every time we upgrade to a new version. I tried to test it and installed an app locally with an older version and then with a newer one many times and most times I didn't get it. Everything was fine. So I'm not sure it's a migration issue but we're able to reproduce it for such flow 😕

**Comment by tanderson-ld** [comment]
@Mik77o , have you ever gotten the error in the non-upgrade case?  Thank you for all the background info.

**Comment by Mik77o** [comment]
> @Mik77o , have you ever gotten the error in the non-upgrade case? Thank you for all the background info.

I will try to check it somehow with QA and come back with an answer as soon as possible. 
From a developer's perspective, I only encountered this problem when I had an app with an older version before and then installed a new one but not always had this bug. I'll see if QA discovers anything.

**Comment by Mik77o** [comment]
@tanderson-ld We tested it and it turns out that even if there was no older version installed before, we get this error. After several restarts of the app, it is possible to move on.

**Comment by tanderson-ld** [comment]
@Mik77o , could you include a code snippet that I can try to use for reproduction of the issue?  Thank you for helping us diagnose this.

**Comment by Mik77o** [comment]
@tanderson-ld 
We do all this logic in the main method.

```
Future<void> main() async {
  await runZonedGuarded(() async {
...
  await _identifyThroughoutServices(keys);
  await di.sl<LaunchDarkly>().startFuture()
...
}
```

```
Future<void> _identifyThroughoutServices(Keys keys) async {
  final launchDarkly = di.sl<LaunchDarkly>();
  final analytics = di.sl<Analytics>();
  final datadogService = di.sl<Datadog>();

  final anonymousId = await analytics.anonymousId ?? '?';

  await launchDarkly.init(
    mobileKey: keys.launchDarklyMobileKey,
    userId: anonymousId,
  );
...
}
```

```
Future<void> init({required String mobileKey, required String userId}) async {
    final config = LDConfigBuilder(mobileKey).build();
    final contextBuilder = LDContextBuilder()..kind('user', userId);
    await LDClient.startWithContext(config, contextBuilder.build());
  }
```

`Future<void> startFuture() => LDClient.startFuture();`

**Comment by panghy** [comment]
@tanderson-ld Thanks for hunting this down (and sorry for the radio silence). We found the issue to be a race condition between us calling setIdentity() before startFuture() was allowed to finish. Funny enough, we also saw `LDClient.get() was called before init()!` but that was because we didn't have a proper context set,

`LDClient.start(config, LDContextBuilder().build())` will throw but a proper LDContextBuilder with at least a single `kind()` call will make the error go away. Hope that helps.
